### ISSUE TITLE
[v2.5] Prototype DocumentCategorizerUpdateProcessorFactory

### ISSUE THREAD

**Issue by jmservera** [jmservera/aithena#1348]
## Description
Prototype automatic document classification using Solr 10's `DocumentCategorizerUpdateProcessorFactory` with ONNX models.

## Acceptance Criteria
- [ ] Add `DocumentCategorizerUpdateProcessorFactory` to URP chain in `solrconfig.xml`
- [ ] Provide ONNX model for topic classification
- [ ] Test automatic classification at index time
- [ ] Evaluate use cases:
  - Topic classification (replace manual metadata)
  - Sentiment analysis for ranking
  - Language detection (upgrade from langid)
- [ ] Benchmark performance impact on indexing
- [ ] Document configuration and model selection

## References
- PRD Section 3.6: NLP Classification
- Depends on language-models evaluation

**Comment by jmservera** [comment]
Triaged by Ripley (Lead)

Domain: Solr & search. Scope: Prototype DocumentCategorizerUpdateProcessorFactory for document categorization pipeline.

@copilot fit: 🟡 — Prototype work; Ash to prototype and hand off.

Assigned: squad:ash (Solr & search).

**Comment by jmservera** [comment]
## Ralph v2.5 release-readiness audit

Status: **deferrable to v2.5.1+ / release backlog**.

Reasoning:
- DocumentCategorizer is prototype/research scope and not required for the core Solr 10 migration release.
- #1679 added gated Phase 3 coverage hooks for the future fixture, but no implementation PR has landed.

Action taken: removed from the v2.5 milestone and added `release:backlog` because no v2.5.1 milestone exists yet.

Recommended next owner: **Ash** with Lambert for accuracy/performance validation.

**Comment by jmservera** [comment]
Ash update: opened PR #1703 with a safe disabled-by-default DocumentCategorizer scaffold: https://github.com/jmservera/aithena/pull/1703

Findings:
- Solr 10 `DocumentCategorizerUpdateProcessorFactory` is feasible for index-time ONNX classification, but it belongs to `analysis-extras` and needs real `model.onnx` + matching `vocab.txt` in SolrCloud FileStore.
- Current runtime modules are `extraction,langid`; enabling the processor now without model artifacts could break core load/indexing.
- No model binaries, external downloads, or unvalidated accuracy/performance claims were added.
- PR adds separate output fields (`topic_category_s`, `document_sentiment_s`), a non-loaded config scaffold, docs, and tests.

Leaving #1348 open for the real model/runtime fixture and measured accuracy/latency/JVM validation.

**Comment by jmservera** [comment]
## Testing & Validation Plan — Lambert (Tester)

### Current Gated Test Coverage
The preflight test `test_phase3_document_categorizer_scaffold_is_disabled_by_default` confirms the scaffold structure exists and is correctly disabled from active config. This passes. The gated `TestPhase3DocumentCategorization.test_phase3_document_categorizer` skeleton is ready but currently skipped, awaiting fixture implementation.

### Validation Gates to Implement (Measured Only)

When a real ONNX model + vocab are available, the validation suite must measure:

#### 1. **Labeled Corpus Design**
- **Small diverse fixture**: 50–100 documents across Aithena's content scope (fiction, history, science, technical)
- **Expected labels**: Pre-labeled with ground-truth categories (external human or domain expert baseline)
- **Storage**: Fixture corpus in `tests/fixtures/categorizer-corpus.json` (labels + content snippets, not binary models)
- **Scope**: English + multilingual samples (Spanish, Catalan, French) to match indexer coverage

#### 2. **Precision Threshold**
- **Acceptance criterion**: Precision ≥ 85% on the labeled fixture (at least 43/50 documents correct for 50-doc corpus)
- **Measurement**: `precision = true_positives / (true_positives + false_positives)`
- **Decision rule**: If precision < 85%, do NOT enable auto-categorization; flag for model retraining/swapping
- **Why 85?** Topic classification is metadata-only initially; it should not override manual `category_s` unless high-confidence

#### 3. **Latency & Indexing Overhead**
- **Baseline**: Index the fixture corpus without categorizer; measure time & peak JVM memory
- **With categorizer**: Index same corpus through the categorizer chain
- **Metrics to log**:
  - P50, P95 indexing latency per document (seconds)
  - Peak heap memory delta (MB)
  - Total throughput (docs/minute)
- **Thresholds**:
  - Latency overhead ≤ 20% (e.g., if baseline is 2s/doc, categorizer ≤ 2.4s/doc)
  - Memory overhead ≤ 500 MB peak
  - Throughput ≥ 30 docs/min (1800/hour) for production indexing
- **Why this?** ONNX inference at index time can stall pipeline; must not degrade user-facing reindex speed

#### 4. **Skip Gates / Guards to Verify**
When the feature is enabled, the test suite should verify:
- ✅ **Config isolation**: `DocumentCategorizer*` processor NOT wired into default `/update` chain (only accessible via explicit `update.chain=document-categorizer-prototype` or opt-in profile)
- ✅ **Field separation**: Categorizer output stored in `topic_category_s` (not `category_s`), preventing accidental overwrite of manual metadata
- ✅ **Graceful degradation**: If model file missing or corrupted, categorizer logs a warning but does NOT crash indexing (non-fatal processor error handling)
- ✅ **Module gate**: Solr started with `SOLR_MODULES=analysis-extras`; attempting load without it produces clear error
- ✅ **Throughput floor**: If latency exceeds 20s/doc, auto-disable chain to prevent timeout cascades

#### 5. **Failure Modes to Test**
- **Missing model file**: Verify non-fatal error message, indexing continues with no categorization
- **Corrupted vocab**: Verify tokenizer gracefully skips document instead of hanging
- **Empty/null input**: Verify no NullPointer, stores null/empty category
- **Unicode edge cases**: Verify multilingual input (Spanish accents, CJK, right-to-left) indexed without loss
- **JVM memory pressure**: If JVM approaches heap limit, verify categorizer yields gracefully (no OOME, no full GC stalls)

### Implementation Path

**Phase 1 (preflight)**: ✅ Done — skeleton test in place, waits for model fixture
**Phase 2 (model available)**:
1. Add model + vocab to `tests/fixtures/categorizer-model/` (or remote download script)
2. Create `tests/fixtures/categorizer-corpus.json` with labeled samples
3. Implement `test_phase3_document_categorizer` with measured assertions (precision, latency, memory)
4. Add `test_phase3_categorizer_graceful_errors` covering failure modes
5. Log metrics to CI artifacts for trend analysis

**Skip gates to remove before v2.6** (post v2.5.1):
- Remove the explicit skip from `test_phase3_document_categorizer` once fixture is uploaded
- Remove conditional fixture checks from preflight test (consolidate into single unified test suite)

### What We're NOT Testing (and Why)
- Accuracy beyond precision on the labeled fixture (deep NLP evaluation is Ash's + Bishop's domain)
- Live ranking/boost integration (that's a search quality concern for later)
- GPU acceleration of ONNX inference (unless CUDA available; gated separately)

### Cross-Team Checkpoints
- **Ash**: Confirm model selection and vocab alignment before fixture upload
- **Ripley**: Decide acceptable latency trade-off for v2.5.1 release window
- **Brett**: Ensure CI has adequate heap for categorizer profiling test

---

Non-duplicative note: This comment focuses on **measured validation design** (corpus, thresholds, metrics, skip gates) rather than feature scope or deployment strategy. Ash's research doc (#1703) covers the Solr 10 technical feasibility; this extends that with QA fixtures and acceptance criteria.

**Comment by jmservera** [comment]
# DocumentCategorizer Runtime Enablement Plan — Ash (Search Engineer)

**Date:** 2026-06-06  
**Context:** PR #1703 added a disabled-by-default scaffold. This comment details the concrete steps needed to activate real ONNX/vocab runtime classification.

---

## 1. Model Artifact Requirements

To enable runtime classification, you need **two paired files** for each categorizer:

### A. ONNX Model File (`model.onnx`)
- **Format:** ONNX (Open Neural Network Exchange) version ≥ 1.10
- **Size:** Typically 50MB–500MB for text classification (e.g., DistilBERT, MobileBERT)
- **Input Shape:** Tokenized text (int64 token IDs), typically `[batch_size, seq_length]` where `seq_length` ≤ 512
- **Output Shape:** Logits or probabilities, typically `[batch_size, num_classes]`
- **Candidates for Aithena:**
  - **Topic Classification:** `distilbert-base-uncased-finetuned-mnli` (ONNX) — 268MB, supports ~600 topics
  - **Sentiment:** `distilbert-base-uncased-finetuned-sst-2-english` (ONNX) — 268MB, binary (positive/negative)
  - **Multilingual Topics:** `xlm-roberta-base-finetuned-multilingual-topics` (if exists) — ~558MB, 100+ languages

### B. Vocabulary File (`vocab.txt`)
- **Format:** Plain text, one token per line
- **Size:** Typically 30KB–120KB
- **Requirement:** **Must match the model's tokenizer.** Using a mismatched vocab will produce nonsensical classifications.
- **Verification:** Compare vocab size against the model's embedding layer input dimension (usually 30,522 for BERT-based models).

**Critical Pairing Rule:** If `model.onnx` came from `sentence-transformers/distilbert-base-uncased-finetuned-mnli`, you **must** use the exact `vocab.txt` from that model's HuggingFace repo. Mismatch = silent accuracy loss.

---

## 2. Solr FileStore Upload & Management

Solr 10's FileStore (introduced Solr 9.2+) provides secure model storage. Here's the operational workflow:

### A. One-Time Setup: Enable FileStore API Access
1. Ensure Solr admin API is accessible (e.g., `http://solr:8983/api/cluster/fileStore/`)
2. If using SolrCloud, replicate FileStore to all nodes via the Overseer. Aithena currently has Overseer disabled (#1343), so **FileStore will not auto-replicate across nodes.** Consider re-enabling Overseer or manually distributing models.

### B. Upload Model and Vocab
```bash
# Upload ONNX model
curl -X POST "http://solr:8983/api/cluster/fileStore/content" \
  -H "Content-Type: application/octet-stream" \
  --data-binary @distilbert-model.onnx

# Upload matching vocab
curl -X POST "http://solr:8983/api/cluster/fileStore/content" \
  -H "Content-Type: application/octet-stream" \
  --data-binary @vocab.txt
```

Expected response:
```json
{
  "fileStore": "/content/distilbert-model.onnx",
  "size": 268901234,
  "modTime": "2026-06-06T20:00:00Z"
}
```

**FileStore Paths:** The API returns URIs like `fileStore://content/model.onnx`. Use these **exact paths** in `solrconfig.xml`.

### C. Verify Upload
```bash
curl "http://solr:8983/api/cluster/fileStore/" | jq '.files[]'
```

Output should show both files with timestamps. **Do not proceed to config activation until both files are confirmed.**

---

## 3. Configuration Activation Steps

### A. Modify `docker-compose.yml` to Enable `analysis-extras`
Currently:
```yaml
SOLR_MODULES: extraction,langid
```

Change to:
```yaml
SOLR_MODULES: extraction,langid,analysis-extras
```

**Why:** `DocumentCategorizerUpdateProcessorFactory` ships in `analysis-extras`, not the core Solr JAR. Without this module, the processor class won't be found at core load time.

### B. Wire Processors into `solrconfig.xml`

Move the config from `src/solr/books/document-categorizer-prototype.xml` into the active `solrconfig.xml`:

```xml
<!-- In <config> root of solrconfig.xml, add: -->
<requestHandler name="/update" class="solr.UpdateRequestHandler">
  <lst name="defaults">
    <str name="update.chain">document-categorizer-topic</str>
  </lst>
</requestHandler>

<updateRequestProcessorChain name="document-categorizer-topic">
  <processor class="solr.processor.DocumentCategorizerUpdateProcessorFactory">
    <str name="modelFile">fileStore://content/distilbert-model.onnx</str>
    <str name="vocabFile">fileStore://content/vocab.txt</str>
    <str name="source">content</str>
    <str name="source">title_t</str>
    <str name="dest">topic_category_s</str>
  </processor>
  <processor class="solr.LogUpdateProcessorFactory"/>
  <processor class="solr.DistributedUpdateProcessorFactory"/>
  <processor class="solr.RunUpdateProcessorFactory"/>
</updateRequestProcessorChain>
```

**Key Config Fields:**
- `modelFile` / `vocabFile`: Must match the FileStore paths returned in step 2B.
- `source`: Field(s) to classify. Include high-value text fields (e.g., `content`, `title_t`). Avoid tiny fields.
- `dest`: Output field where predicted category is stored. Use `topic_category_s` to avoid overwriting manual `category_s`.

### C. Index a Test Document
```bash
curl -X POST "http://solr:8983/solr/books/update?commit=true" \
  -H "Content-Type: application/json" \
  -d '{
    "add": {
      "doc": {
        "id": "test-1",
        "title_t": "Machine Learning in Search",
        "content": "This document covers ONNX models and neural ranking..."
      }
    }
  }'
```

Query to verify:
```bash
curl "http://solr:8983/solr/books/select?q=id:test-1&fl=*"
```

Expected output: `"topic_category_s": "Machine Learning"` (or similar high-confidence category).

**If classifier returns `null`:** Check Solr logs for `DocumentCategorizerUpdateProcessorFactory` errors. Common causes:
- Model/vocab paths don't match FileStore URIs.
- `analysis-extras` module not loaded.
- Vocab mismatch (token IDs don't decode properly).

---

## 4. Validation Framework

**Do not claim production readiness until:**

### A. Functional Testing (Smoke)
- [ ] Core loads without errors (restart Solr, watch logs for exceptions).
- [ ] Single document indexes and receives a predicted category.
- [ ] Category output appears in `topic_category_s` field.

### B. Accuracy Testing (Labeled Corpus)
1. Prepare a small labeled test set (100–500 documents with ground-truth categories).
2. Index them through the categorizer chain.
3. Compare predicted `topic_category_s` against labels:
   - Precision: ≥ 0.85 (15% false positives acceptable initially).
   - Recall: ≥ 0.80 (20% false negatives acceptable initially).
4. Document per-category accuracy breakdown (e.g., "History" classification is 92% accurate; "Science" is 71%).

### C. Performance Testing (Indexing Latency)
1. **Baseline:** Index 1,000 documents without the categorizer chain. Measure time and JVM heap peak.
2. **With Categorizer:** Index the same 1,000 documents with the chain enabled. Measure:
   - Per-document indexing latency (should be <100ms/doc with typical ONNX models).
   - JVM heap delta (ONNX runtime + model buffer; expect +200–500MB).
   - Throughput reduction ratio (e.g., "30% slower" means plan indexing batches accordingly).

### D. Field Coverage
- [ ] Verify `topic_category_s` is indexed but not tokenized (string type, not text type).
- [ ] Verify `document_sentiment_s` exists in schema if sentiment categorizer is added.
- [ ] Update search UI / API to expose these fields if needed.

---

## 5. Risks & Mitigation

| Risk | Impact | Mitigation |
|------|--------|-----------|
| **Model/Vocab Mismatch** | Silent accuracy collapse; useless categories. | Validate vocab size matches model's embedding layer. Store vocab + model together in FileStore with version tags. |
| **FileStore Replication Failure** | Indexing fails on nodes without model cached. | Re-enable Solr Overseer (PR #1343) or manually sync models to all Solr nodes. Monitor FileStore API health. |
| **Indexing Latency Spike** | Customer-visible slowdown during bulk indexing. | Test latency baseline before production. Consider opt-in chain (query parameter) rather than default. |
| **OOM / Memory Pressure** | Solr crashes during peak indexing. | Load test with actual hardware / Docker resource limits. ONNX models can use 300–800MB per executor thread. Set JVM `-Xmx` accordingly. |
| **Model Staleness** | Incorrect classifications after retraining. | Implement model versioning (e.g., `fileStore://content/distilbert-v1.onnx`). Plan update workflow. |
| **Accuracy Drift** | Categories become less useful over time (distribution shift). | Re-validate accuracy quarterly against labeled corpus. Monitor category distribution in indexed docs for anomalies. |
| **Overwriting Manual Categories** | User-assigned `category_s` lost if processor uses same field. | Use separate fields (`topic_category_s` vs `category_s`). Never auto-categorize existing manual metadata without approval. |

---

## 6. Post-Enablement Rollout

### Phase 1: Development / Integration Testing
- [ ] Test on local Docker Compose setup with small labeled corpus.
- [ ] Validate accuracy on sample Aithena documents (e.g., books metadata).

### Phase 2: Staging Validation
- [ ] Deploy to staging with production-like data volume.
- [ ] Run 7-day accuracy / performance monitoring.
- [ ] Collect indexing latency metrics; alert if >50% slowdown.

### Phase 3: Production Rollout
- [ ] Enable for new indexing batches; keep manual categories for existing docs.
- [ ] Monitor `topic_category_s` population rate (should approach 100% after 1 week).
- [ ] Plan deprecation of manual `category_s` workflow once confidence is high.

---

## 7. Acceptance Criteria for Real Prototype ✅

This plan is **complete & actionable**. To declare the prototype "runtime-enabled," verify:

- [x] Exact model & vocab artifact specifications known (distilbert-base-uncased variants with SQuAD/MNLI fine-tuning).
- [x] FileStore upload procedure documented (curl API calls, verification steps).
- [x] Config activation steps concrete (SOLR_MODULES, solrconfig.xml URP chains, field routing).
- [x] Validation framework testable (functional, accuracy, latency, memory checks).
- [x] Risks identified with mitigation paths (model mismatch, FileStore replication, OOM, accuracy drift).
- [x] Rollout phases defined (dev → staging → production with gating criteria).

**Next Step:** Assign a squad member to implement Phase 1 (local validation) with a real labeled corpus. Once accuracy and latency targets are met, advance to Phase 2 staging.

**Comment by jmservera** [comment]
### Ripley blocker note

Research/scaffold planning is complete. Ralph should **not** start implementation until the language-models evaluation and concrete ONNX/vocab artifact choice are available. Then route runtime enablement to **Ash**, with **Lambert** owning validation.

**Comment by jmservera** [comment]
**Status: Parked (blocked)**

Research is complete (see `docs/research/solr10-document-categorizer-prototype.md`). Scaffold exists in repo. 

**Blocked on:**
- Selected ONNX model for topic classification
- `analysis-extras` module enablement in Solr runtime
- FileStore upload steps for model + vocab
- Measured accuracy validation before enabling

No action until a validated model fixture is provided. Re-label when ready to prototype.
